MEN-4384 Management endpoints: set, get, remove key-value pairs.

merlin-northern opened the pull request:
MEN-4384 Management endpoints: set, get, remove key-value pairs.

    * PUT and GET are used to store and retrieve the whole "expected" conf set
    * to remove send empty array

ChangeLog:none
Signed-off-by: Peter Grzybowski <peter@northern.tech>

merlin-northern commented on api/http/router.go:
```diff
@@ -58,6 +60,10 @@ func NewRouter(app app.App) http.Handler {
 	router.Use(accesslog.Middleware())
 	// requestid attaches X-Men-Requestid header to context
 	router.Use(requestid.Middleware())
+	router.Use(identity.Middleware(
+		identity.NewMiddlewareOptions().
+			SetPathRegex(`^/api/management/v[0-9]/`),
+	))
```
believe it or not but on my setup this is required, otherwise the identity is not propagated to the handler. For Alf it works, so if you think I should remove, I will remove it.

tranchitella replied:
@alfrunes ^^^

tranchitella commented on api/http/management.go:
```diff
@@ -27,3 +37,66 @@ func NewManagementAPI(app app.App) *ManagementAPI {
 	}
 }
+
+func (api *ManagementAPI) ConfigurationSet(c *gin.Context) {
```
Not super important, but I would suggest VerbNoun naming, thus `SetConfiguration`

merlin-northern replied:
changed.

tranchitella commented on api/http/management.go:
```diff
@@ -27,3 +37,66 @@ func NewManagementAPI(app app.App) *ManagementAPI {
+	c.Status(http.StatusCreated)
+}
+
+func (api *ManagementAPI) ConfigurationGet(c *gin.Context) {
```
Not super important, but I would suggest VerbNoun naming, thus `GetConfiguration`

merlin-northern replied:
changed.

tranchitella commented on api/http/router.go:
```diff
@@ -38,6 +38,8 @@ const (
 	URITenantDevices = "/tenants/:tenant_id/devices"
 	URITenantDevice  = "/tenants/:tenant_id/devices/:device_id"
 
+	URIConfiguration = "/config/devices/:device_id"
```
I would suggest naming this endpoint `configurations`, to match our naming convention in other services (e.g. `deployments`, `images`, no abbreviation and plural).

merlin-northern replied:
changed.

tranchitella commented on app/app.go:
```diff
@@ -42,6 +42,9 @@ type App interface {
 	ProvisionDevice(ctx context.Context, dev model.NewDevice) error
 	DecommissionDevice(ctx context.Context, devID uuid.UUID) error
+
+	ConfigurationSet(ctx context.Context, devID uuid.UUID, configuration model.Configuration) error
```
Again, `SetConfiguration` is a better name, IMHO.

merlin-northern replied:
changed.

tranchitella commented on docs/management_api.yml:
```diff
@@ -28,83 +28,80 @@ tags:
+                properties:
+                  id:
+                    type: string
+                  expected:
```
About naming: to be honest, I don't like the name `expected`. I would prefer `required`, or maybe `config` or `configuration`. The `expected` term is a bit out of context here.

merlin-northern replied:
expected and actual are the names that I had agreed upon with Manuel before and at that time I do not think this code was merged. I can change that no problem.

merlin-northern replied:
ok, I lied a bit: I have not changed that. do you think we can keep expected/actual? I agreed on it with Manuel.

tranchitella commented on docs/management_api.yml:
```diff
@@ -28,83 +28,80 @@ tags:
+                    type: array
+                    items:
+                      $ref: '#/components/schemas/Parameter'
+                  actual:
```
To make it consistent with the `reported_ts`, I would rename this field `reported`.

merlin-northern replied:
I removed all attributes, and the GET endpoint returns only the key-value pairs and not the whole device.

tranchitella replied:
Looks better now, but you did not update the API docs. This diff makes it:

```
diff --git a/docs/management_api.yml b/docs/management_api.yml
index 125f4c6..1f4b250 100644
--- a/docs/management_api.yml
+++ b/docs/management_api.yml
@@ -28,7 +28,7 @@ tags:
   - name: ManagementAPI

 paths:
-  /config/devices/{deviceId}:
+  /configurations/devices/{deviceId}:
     get:
       operationId: Get key-value pair store
       tags:
@@ -51,10 +51,10 @@ paths:
                 properties:
                   expected:
                     type: object
-                    $ref: '#/components/schemas/Parameter'
+                    $ref: '#/components/schemas/Configuration'
                   actual:
                     type: object
-                    $ref: '#/components/schemas/Parameter'
+                    $ref: '#/components/schemas/Configuration'
         400:
           description: Bad Request.
           content:
@@ -76,7 +76,7 @@ paths:
         content:
           application/json:
             schema:
-              $ref: '#/components/schemas/Parameter'
+              $ref: '#/components/schemas/Configuration'
       parameters:
         - in: path
           name: deviceId
@@ -112,15 +112,9 @@ components:
         The JWT can be alternatively passed as a cookie named "JWT".

   schemas:
-    Parameter:
+    Configuration:
       type: object
-      properties:
-        key:
-          type: string
-          description: Configuration key.
-        value:
-          type: string
-          description: Configuration value.
+      additionalProperties: true

     Error:
       type: object
```

tranchitella commented on docs/management_api.yml:
```diff
@@ -28,83 +28,80 @@ tags:
+                      $ref: '#/components/schemas/Parameter'
+                  updated_ts:
+                    type: string
+                  report_ts:
```
To match `updated_ts`, I would suggest `reported_ts`.

merlin-northern replied:
I removed all attributes, and the GET endpoint returns only the key-value pairs and not the whole device.

tranchitella commented on docs/management_api.yml:
```diff
@@ -28,83 +28,80 @@ tags:
+                    type: array
+                    items:
+                      $ref: '#/components/schemas/Parameter'
+                  updated_ts:
+                    type: string
```
Can we also have `created_ts`?

merlin-northern replied:
I removed all attributes, and the GET endpoint returns only the key-value pairs and not the whole device.

tranchitella replied:
To be honest, I liked the idea of returning created_ts / updated_ts / reported_ts.
Why did you remove them alltogether?

tranchitella commented on docs/management_api.yml:
```diff
@@ -28,83 +28,80 @@ tags:
+                    items:
+                      $ref: '#/components/schemas/Parameter'
+                  actual:
+                    type: array
```
I see you implemented the set/get interface using an array. I would suggest using a free object here, which makes it possible in the future to evolve to full JSON if needed. Instead if sending `[{"key": "a", "value": "b"}]` I would like to be able to send `{"a": "b"}`. This applies both to the GET and to the PUT method. I know this changes your implementation a lot, but it is way more user-friendly and flexible than a list of key/value attributes. By the way, in the current implementation you can provide the same key multiple times providing multiple items with the same key, the latest is the one inserted.

merlin-northern replied:
I know that, and my original agreement with Manuel was to use the exactly what you say. however, what I did was dictated by what we already have in the code:
```
DesiredAttributes []Attribute
```
ok, I will return to the original idea.

merlin-northern replied:
back to the first implementation. in other words: fixed.

tranchitella commented on docs/management_api.yml:
```diff
@@ -28,83 +28,80 @@ tags:
-              oneOf:
-                - $ref: '#/components/schemas/Parameter'
-                - type: array
+              type: object
+              properties:
+                expected:
+                  type: array
```
See my comment above: the interface should allow us to specify a JSON in the form `{"key": "value"}`, not a list of attributes.

merlin-northern replied:
that was original implementation and original conclusion with Manuel. but the existing code does not go in that direction. I will fix.

merlin-northern replied:
fixed.

tranchitella replied:
As mentioned above, fixed in the code, but the API docs is outdated.

tranchitella commented on docs/management_api.yml:
```diff
@@ -28,83 +28,80 @@ tags:
       responses:
-        200:
-          description: Parameters set and updated successfully.
+        201:
+          description: Created
```
201 is not correct, because this method is an upsert. To make it easier to use, though, the PUT should return 200 and the updated device configuration (full object). This simplifies the consumption of the end-point by the UI, for example.

merlin-northern replied:
204 it is.

tranchitella replied:
204 is way better, thanks. I still believe it would be useful to return the updated object, though, instead of an empty response. I won't fight for this, take it as a suggestion.

tranchitella reviewed (changes requested):
Sorry, a lot of requests but it is way easier to handle them now than later on when we merge the first implementation. If it helps you, you can find the acceptance tests here: https://github.com/mendersoftware/deviceconfig/pull/6
You can run them with `make acceptance-tests`.
They pass with your current implementation, but needs to be adapted if you address my concerns above.

merlin-northern commented on api/http/router.go:
```diff
@@ -69,8 +75,12 @@ func NewRouter(app app.App) http.Handler {
-	// mgmtAPI := NewManagementAPI(app)
+	mgmtAPI := NewManagementAPI(app)
 	mgmtGrp := router.Group(URIManagement)
+
+	mgmtGrp.GET(URIConfiguration, mgmtAPI.ConfigurationGet)
+	mgmtGrp.PUT(URIConfiguration, mgmtAPI.ConfigurationSet)
+
 	// identity middleware for collecting JWT claims into request Context.
 	mgmtGrp.Use(identity.Middleware())
```
> @merlin-northern This is why it didn't work. The order matters - `*.Use` must come first, otherwise the middleware is not pushed on the handler stack.

ha! I knew it! thanks!

merlin-northern commented on app/mocks/App.go:
```diff
@@ -12,7 +12,7 @@
 //    limitations under the License.
 
-// Code generated by mockery v2.2.2. DO NOT EDIT.
+// Code generated by mockery v0.0.0-dev. DO NOT EDIT.
```
to be honest I changed the version manually (only tow or three mocks were added, so no harm done). `go generate` uses docker (!) and I got out of disc space. noted for the future. thanks.

merlin-northern commented:
@tranchitella @alfrunes I addressed all comments, left only the expected/actual as it was agreed earlier.

tranchitella commented on docs/management_api.yml:
```diff
@@ -28,83 +28,67 @@ tags:
 
 paths:
-  /configuration:
+  /config/devices/{deviceId}:
```
We also need to update the API docs to match the new URL (/configurations)

tranchitella reviewed (changes requested):
I updated my PR with the acceptance tests: https://github.com/mendersoftware/deviceconfig/pull/6

They pass, but you need to include in your PR this commit because you did not update the API docs: https://github.com/tranchitella/deviceconfig/commit/65803ae785bb76b4e742fdcc20e0619bc04606b8

Another note: with the current implementation I cannot set empty values for configuration keys, and I believe this is wrong: if we apply the "env variable" model, an empty string as value should work just fine.

merlin-northern commented:
> I updated my PR with the acceptance tests: #6
> 
> They pass, but you need to include in your PR this commit because you did not update the API docs: [tranchitella@65803ae](https://github.com/tranchitella/deviceconfig/commit/65803ae785bb76b4e742fdcc20e0619bc04606b8)

right! added.

> 
> Another note: with the current implementation I cannot set empty values for configuration keys, and I believe this is wrong: if we apply the "env variable" model, an empty string as value should work just fine.

I understood that we either set all or remove all. this means removing the key that are set to empty
values, more complicated.

alfrunes commented on store/mongo/datastore_mongo.go:
```diff
@@ -150,6 +150,31 @@ func (db *MongoStore) InsertDevice(ctx context.Context, dev model.Device) error
+		},
+	}
+
+	_, err := collDevs.UpdateOne(ctx, fltr, update, mopts.Update().SetUpsert(true))
+
+	return errors.Wrap(err, "mongo: failed to store device configuration")
```
I don't know if this can be useful, but we can easily return the updated document as well:
```suggestion
	dev := new(model.Device)
	err := collDevs.FindOneAndUpdate(ctx, fltr, update, mopts.
		FindOneAndUpdate().
		SetUpsert(true).
		SetReturnDocument(mopts.After),
	).Decode(dev)
	return dev, errors.Wrap(err, "mongo: failed to store device configuration")
```

merlin-northern replied:
lets just skip that one for now.

alfrunes commented on api/http/management.go:
```diff
@@ -27,3 +35,61 @@ func NewManagementAPI(app app.App) *ManagementAPI {
+		return
+	}
+
+	err = c.ShouldBindJSON(&configuration)
```
There is no validation of the request body here.

merlin-northern replied:
discussed, and actaully you did it.

merlin-northern commented on model/device.go:
```diff
@@ -22,14 +22,19 @@ import (
 	"github.com/pkg/errors"
 )
 
+type Configuration struct {
+	Expected map[string]interface{} `bson:"expected,omitempty" json:"expected"`
+	Actual   map[string]interface{} `bson:"actual,omitempty" json:"actual"`
+}
+
```
sure.

merlin-northern commented on docs/management_api.yml:
```diff
@@ -130,18 +112,9 @@ components:
-            # - type: number
-            # - type: boolean
-            # - type: array
+      additionalProperties: true
```
sure!

merlin-northern commented on docs/management_api.yml:
```diff
@@ -28,83 +28,67 @@ tags:
 
 paths:
-  /configuration:
+  /configurations/devices/{deviceId}:
```
sure.

alfrunes commented on api/http/management.go:
```diff
@@ -27,3 +36,70 @@ func NewManagementAPI(app app.App) *ManagementAPI {
+
+	device, err := api.App.GetDevice(ctx, devID)
+	if err != nil {
+		rest.RenderError(c,
+			http.StatusNotFound,
+			errors.New(http.StatusText(http.StatusNotFound)),
+		)
+		return
```
Sorry I didn't catch this in the first round, but we need to detect if the error is actually internal:
```suggestion
		switch cause := errors.Cause(err); cause {
		case store.ErrDeviceNoExist:
			rest.RenderError(c,
				http.StatusNotFound,
				errors.New(cause),
			)
		default:
			c.Error(err)
			rest.RenderError(c,
				http.StatusNotFound,
				errors.New(http.StatusText(http.StatusInternalServerError)),
			)
		}
		return
```

merlin-northern replied:
fixed

alfrunes commented on model/device.go:
```diff
@@ -27,15 +27,15 @@ type Device struct {
 
 	// DesiredAttributes is the configured attributes for the device.
-	DesiredAttributes []Attribute `bson:"desired,omitempty" json:"desired"`
+	DesiredAttributes Attributes `bson:"expected,omitempty" json:"configured"`
```
```suggestion
	DesiredAttributes Attributes `bson:"configured,omitempty" json:"configured"`
```

merlin-northern replied:
fixed

alfrunes commented on model/device.go:
```diff
@@ -27,15 +27,15 @@ type Device struct {
+	DesiredAttributes Attributes `bson:"expected,omitempty" json:"configured"`
 	// CurrentAttributes is the configuration reported by the device.
-	CurrentAttributes []Attribute `bson:"current,omitempty" json:"current"`
+	CurrentAttributes Attributes `bson:"actual,omitempty" json:"reported"`
```
```suggestion
	CurrentAttributes Attributes `bson:"reported,omitempty" json:"reported"`
```

merlin-northern replied:
fixed

alfrunes reviewed (changes requested):
Nearly there, I just have some smaller changes I didn't catch the first round.

alfrunes commented on store/mongo/datastore_mongo.go:
```diff
@@ -150,6 +150,31 @@ func (db *MongoStore) InsertDevice(ctx context.Context, dev model.Device) error
+				Key:   "expected",
+				Value: dev.DesiredAttributes,
+			},
+		},
```
You're not setting `updated_ts` here.

merlin-northern replied:
added.

alfrunes commented on store/mongo/datastore_mongo.go:
```diff
@@ -164,3 +189,26 @@ func (db *MongoStore) DeleteDevice(ctx context.Context, devID uuid.UUID) error {
+
+	var device model.Device
+
+	if res == nil {
+		return device, errors.Wrap(store.ErrDeviceNoExist, "mongo")
+	}
```
This will always evaluate to false.

merlin-northern replied:
removed

alfrunes commented on api/http/management.go:
```diff
@@ -27,3 +37,79 @@ func NewManagementAPI(app app.App) *ManagementAPI {
+		case store.ErrDeviceNoExist:
+			rest.RenderError(c,
+				http.StatusNotFound,
+				errors.New(http.StatusText(http.StatusNotFound)),
+			)
+			return
+		default:
+			rest.RenderError(c,
```
```suggestion
				cause,
			)
			return
		default:
			c.Error(err)
			rest.RenderError(c,
```

merlin-northern replied:
sure

alfrunes commented on api/http/management.go:
```diff
@@ -27,3 +37,81 @@ func NewManagementAPI(app app.App) *ManagementAPI {
+			rest.RenderError(c,
+				http.StatusBadRequest,
+				errors.Wrap(err, "invalid request body"),
+			)
```
```suggestion
			)
			return
```

merlin-northern replied:
sure

tranchitella commented:
Failure in integration test is unrelated, let's merge.
